src/utils.jl

```diff
@@ -313,50 +202,104 @@ function split_dataset(data::Matrix; train_ratio::Float64 = 0.9, randomize::Bool
+_haschoitags(M::Union{MPS,MPO}) = _haschoitags(LPDO(M)) 
 
-function ischoi(M::MPO)
-  return (length(inds(M[1],"Site")) == 4 ? true : false)
-  #return ( length(inds(M[1])) == 5 ? true : false)
-end
 
-function makeUnitary(L::LPDO{MPS})
-  ψ = L.X
-  U = MPO(ITensor[copy(ψ[j]) for j in 1:length(ψ)])
-  prime!(U,tags="Output")
-  removetags!(U, "Input")
-  removetags!(U, "Output")
-  return U
+"""
+    _choitags(U::Union{MPS, MPO})
+
+Assign the input/output tags defined for a Choi matrix to an MPO.
+
+  σ₁ -o- σ₁′       σ₁ⁱ -o- σ₁ᴼ   
+      |                 |
+  σ₂ -o- σ₂′  ⟶    σ₂ⁱ -o- σ₂ᴼ
+      |                 |
+  σ₃ -o- σ₃′       σ₃ⁱ -o- σ₃ᴼ
+                  
+"""
+function _choitags(L::LPDO{MPO})
+  _haschoitags(L) && return L
+  U = copy(L.X)
+  U = addtags(U, "Input", plev = 0, tags = "Qubit")
+  U = addtags(U, "Output", plev = 1, tags = "Qubit")
+  return LPDO(noprime(U))
 end
 
-function makeChoi(U0::MPO)
-  M = MPS(ITensor[copy(U0[j]) for j in 1:length(U0)])
-  addtags!(M, "Input", plev = 0, tags = "Qubit")
-  addtags!(M, "Output", plev = 1, tags = "Qubit")
-  noprime!(M)
-  return LPDO(M)
+_choitags(U::MPO) = _choitags(LPDO(U)).X
+
+"""
+    _mpotags(Ψ::MPS)
+
+Inverse of _choitags
+"""
+
+function _mpotags(L::LPDO)
+  !_haschoitags(L) && return L
+  U = copy(L.X) 
+  U = prime(U,tags="Output")
+  U = removetags(U, "Input")
+  U = removetags(U, "Output")
+  return LPDO(U)
 end
 
+_mpotags(M::Union{MPS,MPO}) = _mpotags(LPDO(M)).X
 
-function numberofqubits(gate::Tuple)
-  s = gate[2]
+
+"""
+    _unitaryMPO_to_choiMPS(L::LPDO{MPO})
+    _unitaryMPO_to_choiMPS(U::MPO)
+
+
+     MPO          MPS (vectorized MPO) 
+  σ₁ -o- σ₁′       o= (σ₁ⁱ,σ₁ᴼ)   
+      |            | 
+  σ₂ -o- σ₂′  ⟶    o= (σ₂ⁱ,σ₂′ᴼ)
+      |            | 
+  σ₃ -o- σ₃′       o= (σ₃ⁱ,σ₃′ᴼ)
+                  
+Transforms a unitary MPO into a Choi MPS with appropriate tags.
+"""
+#_unitaryMPO_to_choiMPS(L::LPDO{MPO}) = LPDO(convert(MPS, _choitags(L).X)) 
+_unitaryMPO_to_choiMPS(U::MPO) = convert(MPS, _choitags(U))
+_unitaryMPO_to_choiMPS(L::LPDO{MPO}) = LPDO(_unitaryMPO_to_choiMPS(L.X))
+
+"""
+    _choiMPS_to_unitaryMPO(L::LPDO{MPS})
+    _choiMPS_to_unitaryMPO(Ψ::MPS)
+
+Transforms a Choi MPS into an MPO with appropriate tags.
+Inverse of `_unitaryMPO_to_choiMPS`.
+"""
+#_choiMPS_to_unitaryMPO(L::LPDO{MPS}) = LPDO(convert(MPO, _mpotags(L).X))
+#_choiMPS_to_unitaryMPO(Ψ::MPS)       = _choiMPS_to_unitaryMPO(LPDO(Ψ)).X
+_choiMPS_to_unitaryMPO(Ψ::MPS)       = convert(MPO, _mpotags(Ψ))
+_choiMPS_to_unitaryMPO(L::LPDO{MPS}) = LPDO(_choiMPS_to_unitaryMPO(L.X)) 
+
+function numberofqubits(g::Tuple)
+  s = g[2]
   n = (s isa Number ? s : maximum(s))
   return n
 end
 
-function numberofqubits(gates::Vector{<:Tuple})
-  nMax = 0
-  for g in gates
-    s = g[2]
-    n = (s isa Number ? s : maximum(s))
-    nMax = (n > nMax ? n : nMax)
-  end
-  return nMax
-end
-
-
+numberofqubits(gates::Vector{<:Any}) =  
+  maximum([numberofqubits(gates[n]) for n in 1:length(gates)])
```

Comment:
Maybe just `nqubits`?

Also for your own knowledge, you could write it as:
```julia
nqubits(gates::Vector) = maximum((nqubits(gate) for gate in gates))
```
Note that if you put it in parentheses instead of `[...]`, it doesn't actually create a vector of all of the number of qubits, just a lazy object that will compute them as needed. A nice efficiency trick.